src/Number/Quaternion.php

```
@@ -0,0 +1,345 @@
+<?php namespace MathPHP\Number;
+
+use MathPHP\Exception;
+use MathPHP\Functions\Special;
+
+/**
+ * Quaternionic Numbers
+ *
+ * A quaternion is a number that can be expressed in the form a + bi + cj + dk,
+ * where a, b, c, andd are real numbers and i, j, and k are the basic quaternions, satisfying the
+ * equation i² = j² = k² = ijk = −1.
+ * https://en.wikipedia.org/wiki/Quaternion
+ */
+class Quaternion implements ObjectArithmetic
+{
+    /**
+     * Real part of the quaternionic number
+     * @var number
+     */
+    protected $r;
+
+    /**
+     * First Imaginary part of the quaternionic number
+     * @var number
+     */
+    protected $i;
+
+    /**
+     * Second Imaginary part of the quaternionic number
+     * @var number
+     */
+    protected $j;
+
+    /**
+     * Third Imaginary part of the quaternionic number
+     * @var number
+     */
+    protected $k;
+
+    /**
+     * Floating-point range near zero to consider insignificant.
+     */
+    const EPSILON = 1e-6;
+
+    /**
+     * Constructor
+     *
+     * @param number $r Real part
+     * @param number $i Imaginary part
+     * @param number $j Imaginary part
+     * @param number $k Imaginary part
+     */
+    public function __construct($r, $i, $j, $k)
+    {
+        $this->r = $r;

```

We probably want to have some sort of check that you don't pass in strings or arrays or something. You want to fail early during construction rather than later on when you do some arithmetic operation which couldn't possibly work. Either add some check (`is_numeric` or `is_int || is_float`) or add a type hint for float in the constructor.

I also noticed the Complex class also has this problem.

BTW, PHP 8 has union types. So in the future we can put `int|float` in the function declaration.